src/java/org/apache/cassandra/audit/AuditLogManager.java

```diff
@@ -0,0 +1,323 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.audit;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.CQLStatement;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.statements.ParsedStatement;
+import org.apache.cassandra.db.fullquerylog.FullQueryLogger;
+import org.apache.cassandra.exceptions.AuthenticationException;
+import org.apache.cassandra.exceptions.UnauthorizedException;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.service.QueryState;
+import org.apache.cassandra.utils.FBUtilities;
+
+public class AuditLogManager
+{
+    static final Logger logger = LoggerFactory.getLogger(AuditLogManager.class);
+    private static final AuditLogManager instance = new AuditLogManager();
+    private IAuditLogger auditLogger;
+    private AuditLogManager()
+    {
+        if (isAuditingEnabled())
+        {
+            logger.info("Audit logging is enabled.");
+            this.auditLogger = getAuditLogger();
+        }
+        else
+        {
+            logger.info("Audit logging is disabled.");
+        }
+    }
+
+    public static AuditLogManager getInstance()
+    {
+        return instance;
+    }
+
+
+    private IAuditLogger getAuditLogger()
+    {
+        String loggerClassName = DatabaseDescriptor.getAuditLoggingOptions().logger;
+        if(loggerClassName !=null)
+        {
+            return FBUtilities.newAuditLogger(loggerClassName);
+        }
+        return FBUtilities.newAuditLogger(BinAuditLogger.class.getName());
+    }
+
+    @VisibleForTesting
+    public IAuditLogger getLogger()
+    {
+        return auditLogger;
+    }
+
+    public boolean isAuditingEnabled()
+    {
+        return DatabaseDescriptor.getAuditLoggingOptions().enabled;
+    }
+    public boolean isLoggingEnabled()
+    {
+        return isAuditingEnabled() || FullQueryLogger.instance.enabled();
+    }
+    public boolean isFQLEnabled()
+    {
+        return FullQueryLogger.instance.enabled();
+    }
+
+    private boolean isSystemKeyspace(String keyspaceName)
+    {
+        return SchemaConstants.isLocalSystemKeyspace(keyspaceName);
+    }
+
+    /**
+     * Logging overloads
+     */
+    public void log(AuditLogEntry logEntry)
+    {
+        if (isAuditingEnabled()
+            && (null != logEntry)
+            && ((null == logEntry.getKeyspace()) || !isSystemKeyspace(logEntry.getKeyspace()))
```

Comment:
shouldn't we audit system keyspaces as well ?

Reply:
System keyspaces are mostly accessed through C* internally.

Reply:
is there a way to figure out that query on system keyspces are internal or external?

Reply:
I am not aware of that.